cpp/src/community/detail/refine.cuh

```diff
@@ -0,0 +1,763 @@
+                       singleton_and_connected_flags.end(),
+                       [] __device__(auto flag) { return flag > 0; });
+
+    if (GraphViewType::is_multi_gpu) {
+      nr_remaining_active_vertices = host_scalar_allreduce(handle.get_comms(),
+                                                           nr_remaining_active_vertices,
+                                                           raft::comms::op_t::SUM,
+                                                           handle.get_stream());
+    }
+
+    if (nr_remaining_active_vertices == 0) { break; }
+
+    // Update Leiden assignment to edge sources and destinitions
+    // and singleton mask to edge sources
+
+    if constexpr (GraphViewType::is_multi_gpu) {
+      src_leiden_assignment_cache =
+        edge_src_property_t<GraphViewType, vertex_t>(handle, graph_view);
+      dst_leiden_assignment_cache =
+        edge_dst_property_t<GraphViewType, vertex_t>(handle, graph_view);
+      src_active_flag_cache = edge_src_property_t<GraphViewType, uint8_t>(handle, graph_view);
+    }
+
+    if constexpr (GraphViewType::is_multi_gpu) {
+      update_edge_src_property(
+        handle, graph_view, leiden_assignment.begin(), src_leiden_assignment_cache);
+
+      update_edge_dst_property(
+        handle, graph_view, leiden_assignment.begin(), dst_leiden_assignment_cache);
+
+      update_edge_src_property(
+        handle, graph_view, singleton_and_connected_flags.begin(), src_active_flag_cache);
+    }
+
+    auto src_input_property_values =
+      GraphViewType::is_multi_gpu
+        ? view_concat(src_louvain_assignment_cache.view(), src_leiden_assignment_cache.view())
+        : view_concat(detail::edge_major_property_view_t<vertex_t, vertex_t const*>(
+                        louvain_assignment.data()),
+                      detail::edge_major_property_view_t<vertex_t, vertex_t const*>(
+                        leiden_assignment.data()));
+
+    auto dst_input_property_values =
+      GraphViewType::is_multi_gpu
+        ? view_concat(dst_louvain_assignment_cache.view(), dst_leiden_assignment_cache.view())
+        : view_concat(detail::edge_minor_property_view_t<vertex_t, vertex_t const*>(
+                        louvain_assignment.data(), vertex_t{0}),
+                      detail::edge_minor_property_view_t<vertex_t, vertex_t const*>(
+                        leiden_assignment.data(), vertex_t{0}));
+
+    rmm::device_uvector<vertex_t> leiden_keys_used_in_edge_reduction(0, handle.get_stream());
+    rmm::device_uvector<weight_t> refined_community_volumes(0, handle.get_stream());
+    rmm::device_uvector<weight_t> refined_community_cuts(0, handle.get_stream());
+
+    //
+    // For each refined community, compute its volume
+    // (i.e.sum of weighted degree of all vertices inside it, ||Cr||) and
+    // and cut between itself and its Louvain community (E(Cr, S-Cr))
+    //
+    // TODO: Can we update ||Cr|| and E(Cr, S-Cr) instead of recomputing?
```

Comment:
Our convention is to use FIXME instead of TODO. We search for FIXME to find/track FIXMEs.